pyfarm/agent/http/api/software.py

```diff
@@ -0,0 +1,204 @@
+    from http.client import OK, ACCEPTED, INTERNAL_SERVER_ERROR, NO_CONTENT
+
+from twisted.web.server import NOT_DONE_YET
+from twisted.internet import reactor
+from twisted.internet.defer import Deferred, inlineCallbacks
+
+import treq
+
+from voluptuous import Schema, Required
+
+from pyfarm.agent.config import config
+from pyfarm.agent.utility import STRINGS
+from pyfarm.agent.logger import getLogger
+from pyfarm.agent.sysinfo.software import check_software_availability
+from pyfarm.agent.http.core.client import (
+    post_direct, delete_direct, http_retry_delay)
+from pyfarm.agent.http.api.base import APIResource
+
+logger = getLogger("agent.http.software")
+
+
+class CheckSoftware(APIResource):
+    """
+    Requests the agent to check whether a given software version is installed
+    locally.
+
+    The agent will asynchronously update its software information on the master.
+
+    .. http:post:: /api/v1/check_software HTTP/1.1
+
+        **Request**
+
+        .. sourcecode:: http
+
+            POST /api/v1/check_software HTTP/1.1
+            Accept: application/json
+
+            {
+                "software": "Blender",
+                "version": "2.72"
+            }
+
+        **Response**
+
+        .. sourcecode:: http
+
+            HTTP/1.1 200 ACCEPTED
+            Content-Type: application/json
+
+    """
+    SCHEMAS = {
+        "POST": Schema({
+            Required("software"): STRINGS,
+            Required("version"): STRINGS})}
+    isLeaf = False  # this is not really a collection of things
+
+    def post(self, **kwargs):
+        request = kwargs["request"]
+        data = kwargs["data"]
+
+        logger.info("Checking whether software %s, version %s is present",
+                    data["software"], data["version"])
+
+        deferred = check_software_availability(data["software"],
+                                               data["version"])
+
+        @inlineCallbacks
+        def mark_software_available(software, version):
+            url = "{master_api}/agents/{agent}/software/".format(
+                master_api=config.get("master_api"),
+                agent=config.get("agent_id"))
+            while True:
+                try:
+                    response = yield post_direct(url, data={
+                            "software": software,
+                            "version": version})
+                except Exception as error:
+                    delay = http_retry_delay()
+                    logger.error(
+                        "Failed to post availability of software %s, "
+                        "version %s to master: %r. Will retry in %s "
+                        "seconds.",
+                        software, version, error, delay)
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                else:
+                    data = yield treq.json_content(response)
```

Comment:
Not really necessary at this point. We're not actually doing anything with the data, so there's no need to parse the json content. We're really just looking at the HTTP status code.